File: js/src/algo_gen/dsl.js
/**
 * dsl.js — Whitelist DSL expression parser for AlgoManifest entry/exit conditions.
 *
 * Recursive-descent parser (no eval, no AST stdlib). Mirrors Python dsl.py.
 *
 * Grammar:
 *   expr     ::= or_expr
 *   or_expr  ::= and_expr ('or' and_expr)*
 *   and_expr ::= not_expr ('and' not_expr)*
 *   not_expr ::= 'not' not_expr | cmp_expr
 *   cmp_expr ::= add_expr (('<'|'<='|'>'|'>='|'=='|'!=') add_expr)*
 *   add_expr ::= mul_expr (('+'|'-') mul_expr)*
 *   mul_expr ::= unary (('*'|'/') unary)*
 *   unary    ::= '-' unary | atom
 *   atom     ::= number | 'true' | 'false' | name | 'pattern' '.' name | '(' expr ')'
 *
 * Public API:
 *   compileExpr(expr, indicatorIds) -> (env) => boolean
 */

// ── Whitelist sets ────────────────────────────────────────────────────────────

export const BAR_FIELDS = new Set([
    'close', 'open', 'high', 'low', 'volume', 'spread',
]);

export const PATTERN_NAMES = new Set([
    'doji', 'hammer', 'engulfing', 'marubozu', 'pin_bar',
    'morning_star', 'evening_star',
    'three_white_soldiers', 'three_black_crows',
    'double_top', 'double_bottom',
    'ascending_triangle', 'descending_triangle',
    'bullish_flag', 'bearish_flag',
    'head_and_shoulders', 'inverse_head_and_shoulders',
    'gartley_bull', 'gartley_bear',
    'bat_bull', 'butterfly_bull', 'crab_bull',
]);

// ── Tokenizer ─────────────────────────────────────────────────────────────────

const TOKEN_NUM   = 'NUM';
const TOKEN_NAME  = 'NAME';
const TOKEN_OP    = 'OP';
const TOKEN_EOF   = 'EOF';

function tokenize(src) {
    const tokens = [];
    let i = 0;
    while (i < src.length) {
        // Skip whitespace
        if (/\s/.test(src[i])) { i++; continue; }
        // Numbers (int or float)
        if (/[0-9]/.test(src[i]) || (src[i] === '.' && /[0-9]/.test(src[i + 1] || ''))) {
            let j = i;
            while (j < src.length && /[0-9]/.test(src[j])) j++;
            if (j < src.length && src[j] === '.') {
                j++;
                while (j < src.length && /[0-9]/.test(src[j])) j++;
            }
            // Scientific notation
            if (j < src.length && (src[j] === 'e' || src[j] === 'E')) {
                j++;
                if (j < src.length && (src[j] === '+' || src[j] === '-')) j++;
                while (j < src.length && /[0-9]/.test(src[j])) j++;
            }
            tokens.push({ type: TOKEN_NUM, value: parseFloat(src.slice(i, j)), pos: i });
            i = j;
            continue;
        }
        // Identifiers / keywords
        if (/[a-zA-Z_]/.test(src[i])) {
            let j = i;
            while (j < src.length && /[a-zA-Z0-9_]/.test(src[j])) j++;
            tokens.push({ type: TOKEN_NAME, value: src.slice(i, j), pos: i });
            i = j;
            continue;
        }
        // Two-char operators
        const two = src.slice(i, i + 2);
        if (['<=', '>=', '==', '!='].includes(two)) {
            tokens.push({ type: TOKEN_OP, value: two, pos: i });
            i += 2;
            continue;
        }
        // Single-char operators / punctuation
        if ('<>+-*/().'.includes(src[i])) {
            tokens.push({ type: TOKEN_OP, value: src[i], pos: i });
            i++;
            continue;
        }
        throw new Error(`DSL: unexpected character '${src[i]}' at position ${i}`);
    }
    tokens.push({ type: TOKEN_EOF, value: null, pos: src.length });
    return tokens;
}

// ── Parser ────────────────────────────────────────────────────────────────────

class Parser {
    constructor(tokens, allowedNames) {
        this._tokens = tokens;
        this._pos    = 0;
        this._allowed = allowedNames;
    }

    _peek() {
        return this._tokens[this._pos];
    }

    _consume() {
        return this._tokens[this._pos++];
    }

    _expect(type, value) {
        const tok = this._consume();
        if (tok.type !== type || (value !== undefined && tok.value !== value)) {
            throw new Error(
                `DSL parse error: expected ${value !== undefined ? `'${value}'` : type} ` +
                `at pos ${tok.pos}, got '${tok.value}'`
            );
        }
        return tok;
    }

    parseExpr() {
        const node = this.parseOrExpr();
        if (this._peek().type !== TOKEN_EOF) {
            const tok = this._peek();
            throw new Error(`DSL parse error: unexpected token '${tok.value}' at pos ${tok.pos}`);
        }
        return node;
    }

    parseOrExpr() {
        let left = this.parseAndExpr();
        while (this._peek().type === TOKEN_NAME && this._peek().value === 'or') {
            this._consume();
            const right = this.parseAndExpr();
            const l = left, r = right;
            left = env => l(env) || r(env);
        }
        return left;
    }

    parseAndExpr() {
        let left = this.parseNotExpr();
        while (this._peek().type === TOKEN_NAME && this._peek().value === 'and') {
            this._consume();
            const right = this.parseNotExpr();
            const l = left, r = right;
            left = env => l(env) && r(env);
        }
        return left;
    }

    parseNotExpr() {
        if (this._peek().type === TOKEN_NAME && this._peek().value === 'not') {
            this._consume();
            const inner = this.parseNotExpr();
            return env => !inner(env);
        }
        return this.parseCmpExpr();
    }

    parseCmpExpr() {
        let left = this.parseAddExpr();
        const CMP_OPS = new Set(['<', '<=', '>', '>=', '==', '!=']);
        while (this._peek().type === TOKEN_OP && CMP_OPS.has(this._peek().value)) {
            const op    = this._consume().value;
            const right = this.parseAddExpr();
            const l = left, r = right;
            switch (op) {
                case '<':  left = env => l(env) <  r(env); break;
                case '<=': left = env => l(env) <= r(env); break;
                case '>':  left = env => l(env) >  r(env); break;
                case '>=': left = env => l(env) >= r(env); break;
                case '==': left = env => l(env) === r(env); break;
                case '!=': left = env => l(env) !== r(env); break;
            }
        }
        return left;
    }

    parseAddExpr() {
        let left = this.parseMulExpr();
        while (
            this._peek().type === TOKEN_OP &&
            (this._peek().value === '+' || this._peek().value === '-')
        ) {
            const op    = this._consume().value;
            const right = this.parseMulExpr();
            const l = left, r = right;
            if (op === '+') left = env => l(env) + r(env);
            else            left = env => l(env) - r(env);
        }
        return left;
    }

    parseMulExpr() {
        let left = this.parseUnary();
        while (
            this._peek().type === TOKEN_OP &&
            (this._peek().value === '*' || this._peek().value === '/')
        ) {
            const op    = this._consume().value;
            const right = this.parseUnary();
            const l = left, r = right;
            if (op === '*') left = env => l(env) * r(env);
            else            left = env => r(env) !== 0 ? l(env) / r(env) : NaN;
        }
        return left;
    }

    parseUnary() {
        if (this._peek().type === TOKEN_OP && this._peek().value === '-') {
            this._consume();
            const inner = this.parseUnary();
            return env => -inner(env);
        }
        return this.parseAtom();
    }

    parseAtom() {
        const tok = this._peek();

        // Parenthesised expression
        if (tok.type === TOKEN_OP && tok.value === '(') {
            this._consume();
            const inner = this.parseOrExpr();
            this._expect(TOKEN_OP, ')');
            return inner;
        }

        // Number literal
        if (tok.type === TOKEN_NUM) {
            this._consume();
            const v = tok.value;
            return () => v;
        }

        // Identifier
        if (tok.type === TOKEN_NAME) {
            const name = tok.value;

            // Boolean literals
            if (name === 'true')  { this._consume(); return () => true; }
            if (name === 'false') { this._consume(); return () => false; }

            // Keywords not valid as identifiers here
            if (name === 'and' || name === 'or' || name === 'not') {
                throw new Error(`DSL parse error: unexpected keyword '${name}' at pos ${tok.pos}`);
            }

            // pattern.X attribute access
            if (name === 'pattern') {
                this._consume();
                this._expect(TOKEN_OP, '.');
                const attrTok = this._consume();
                if (attrTok.type !== TOKEN_NAME) {
                    throw new Error(`DSL parse error: expected pattern attribute name at pos ${attrTok.pos}`);
                }
                const attr = attrTok.value;
                if (!PATTERN_NAMES.has(attr)) {
                    throw new Error(
                        `DSL: pattern.${attr} is not a known pattern name; ` +
                        `allowed: ${[...PATTERN_NAMES].sort().join(', ')}`
                    );
                }
                return env => {
                    const p = env['pattern'];
                    if (!p) return false;
                    const v = typeof p[attr] !== 'undefined' ? p[attr] : false;
                    return Boolean(v);
                };
            }

            // Must be in allowed names
            if (!this._allowed.has(name)) {
                throw new Error(
                    `DSL: Name '${name}' is not declared; ` +
                    `allowed names: ${[...this._allowed].sort().join(', ')}`
                );
            }
            this._consume();
            return env => {
                const v = env[name];
                // null/undefined during warmup → treat as 0 (falsy)
                if (v === null || v === undefined || (typeof v === 'number' && isNaN(v))) return null;
                return v;
            };
        }

        throw new Error(`DSL parse error: unexpected token '${tok.value}' at pos ${tok.pos}`);
    }
}

// ── Public API ────────────────────────────────────────────────────────────────

/**
 * Compile a DSL expression string into an evaluator function.
 *
 * @param {string}      expr          expression string
 * @param {Set<string>} indicatorIds  declared indicator IDs from the manifest
 * @returns {(env: object) => boolean}  evaluator function
 * @throws {Error}  on syntax or whitelist error
 */
export function compileExpr(expr, indicatorIds) {
    if (!expr || !expr.trim()) {
        throw new Error('DSL expression must be non-empty');
    }
    const allowedNames = new Set([...indicatorIds, ...BAR_FIELDS, 'pattern']);
    const tokens = tokenize(expr.trim());
    const parser = new Parser(tokens, allowedNames);
    const fn     = parser.parseExpr();
    // Wrap to ensure boolean return and handle null propagation
    return env => {
        try {
            const v = fn(env);
            if (v === null || v === undefined) return false;
            return Boolean(v);
        } catch (_) {
            return false;
        }
    };
}

File: cpp/src/algo_gen/dsl.cpp
/**
 * AlgoForge — src/algo_gen/dsl.cpp
 * S1 Phase 1 — Whitelist DSL expression parser (recursive descent).
 *
 * Grammar (mirrors Python dsl.py semantics):
 *   expr   = or_expr
 *   or_expr  = and_expr ('or'  and_expr)*
 *   and_expr = not_expr ('and' not_expr)*
 *   not_expr = 'not'? cmp_expr
 *   cmp_expr = add_expr (('<'|'<='|'>'|'>='|'=='|'!=') add_expr)?
 *   add_expr = mul_expr (('+' | '-') mul_expr)*
 *   mul_expr = unary (('*' | '/') unary)*
 *   unary    = '-'? primary
 *   primary  = NUMBER | NAME | 'pattern' '.' NAME | '(' expr ')'
 *
 * Allowed NAMEs:
 *   bar fields: close open high low volume spread
 *   declared indicator IDs
 *   special: pattern (prefix for pattern.X access)
 *
 * pattern.X is an attribute access producing a bool.
 */

#include "algo_gen_internal.hpp"

#include <cctype>
#include <cstdlib>
#include <memory>
#include <set>
#include <stdexcept>
#include <string>
#include <cmath>

namespace algoforge::algo_gen::dsl {

/* =========================================================================
 * Known names
 * ========================================================================= */

static const std::set<std::string> BAR_FIELDS = {
    "close","open","high","low","volume","spread"
};

static const std::set<std::string> PATTERN_NAMES = {
    "doji","hammer","engulfing","marubozu","pin_bar",
    "morning_star","evening_star",
    "three_white_soldiers","three_black_crows",
    "double_top","double_bottom",
    "ascending_triangle","descending_triangle",
    "bullish_flag","bearish_flag",
    "head_and_shoulders","inverse_head_and_shoulders",
    "gartley_bull","gartley_bear",
    "bat_bull","butterfly_bull","crab_bull",
};

/* =========================================================================
 * Tokeniser
 * ========================================================================= */

struct Token {
    enum Kind {
        Num, Id, LParen, RParen, Dot,
        Plus, Minus, Star, Slash,
        Lt, LtEq, Gt, GtEq, EqEq, NEq,
        And, Or, Not,
        Eof, Unknown
    } kind;
    std::string text;
    double      num = 0.0;
};

struct Lexer {
    const char* src;
    size_t      len;
    size_t      pos = 0;

    explicit Lexer(const std::string& s) : src(s.data()), len(s.size()) {}

    void skip_ws() {
        while (pos < len && std::isspace((unsigned char)src[pos])) ++pos;
    }

    Token next() {
        skip_ws();
        if (pos >= len) return {Token::Eof, ""};
        char c = src[pos];

        // Numbers
        if (std::isdigit((unsigned char)c) || (c == '.' && pos+1 < len && std::isdigit((unsigned char)src[pos+1]))) {
            size_t start = pos;
            while (pos < len && (std::isdigit((unsigned char)src[pos]) || src[pos] == '.')) ++pos;
            if (pos < len && (src[pos] == 'e' || src[pos] == 'E')) {
                ++pos;
                if (pos < len && (src[pos] == '+' || src[pos] == '-')) ++pos;
                while (pos < len && std::isdigit((unsigned char)src[pos])) ++pos;
            }
            std::string s(src + start, pos - start);
            Token t;
            t.kind = Token::Num;
            t.text = s;
            t.num  = std::stod(s);
            return t;
        }

        // Identifiers / keywords
        if (std::isalpha((unsigned char)c) || c == '_') {
            size_t start = pos;
            while (pos < len && (std::isalnum((unsigned char)src[pos]) || src[pos] == '_')) ++pos;
            std::string s(src + start, pos - start);
            if (s == "and") return {Token::And, s};
            if (s == "or")  return {Token::Or,  s};
            if (s == "not") return {Token::Not, s};
            return {Token::Id, s};
        }

        ++pos;
        switch (c) {
            case '(': return {Token::LParen,  "("};
            case ')': return {Token::RParen,  ")"};
            case '.': return {Token::Dot,     "."};
            case '+': return {Token::Plus,    "+"};
            case '-': return {Token::Minus,   "-"};
            case '*': return {Token::Star,    "*"};
            case '/': return {Token::Slash,   "/"};
            case '<':
                if (pos < len && src[pos] == '=') { ++pos; return {Token::LtEq, "<="}; }
                return {Token::Lt, "<"};
            case '>':
                if (pos < len && src[pos] == '=') { ++pos; return {Token::GtEq, ">="}; }
                return {Token::Gt, ">"};
            case '=':
                if (pos < len && src[pos] == '=') { ++pos; return {Token::EqEq, "=="}; }
                return {Token::Unknown, "="};
            case '!':
                if (pos < len && src[pos] == '=') { ++pos; return {Token::NEq, "!="}; }
                return {Token::Unknown, "!"};
            default:
                return {Token::Unknown, std::string(1,c)};
        }
    }

    // Peek ahead without consuming
    Token peek() {
        size_t saved = pos;
        Token t = next();
        pos = saved;
        return t;
    }
};

/* =========================================================================
 * Recursive-descent parser
 * ========================================================================= */

struct Parser {
    Lexer lexer;
    Token cur;
    const std::vector<std::string>& ind_ids;
    std::set<std::string> allowed_names;

    Parser(const std::string& src, const std::vector<std::string>& ids)
        : lexer(src), ind_ids(ids)
    {
        for (const auto& id : ids) allowed_names.insert(id);
        for (const auto& bf : BAR_FIELDS) allowed_names.insert(bf);
        allowed_names.insert("pattern");
        cur = lexer.next();
    }

    Token consume() {
        Token t = cur;
        cur = lexer.next();
        return t;
    }

    bool at(Token::Kind k) const { return cur.kind == k; }

    Expr make_binop(const std::string& op, Expr l, Expr r) {
        auto n = std::make_shared<Node>();
        n->kind = NodeKind::BinOp;
        n->name = op;
        n->children = {l, r};
        return n;
    }
    Expr make_cmpop(const std::string& op, Expr l, Expr r) {
        auto n = std::make_shared<Node>();
        n->kind = NodeKind::CmpOp;
        n->name = op;
        n->children = {l, r};
        return n;
    }
    Expr make_bool(NodeKind k, Expr l, Expr r) {
        auto n = std::make_shared<Node>();
        n->kind = k;
        n->children = {l, r};
        return n;
    }

    // Parse full expression
    Expr parse_expr() {
        return parse_or();
    }

    Expr parse_or() {
        Expr left = parse_and();
        while (at(Token::Or)) {
            consume();
            Expr right = parse_and();
            left = make_bool(NodeKind::BoolOr, left, right);
        }
        return left;
    }

    Expr parse_and() {
        Expr left = parse_not();
        while (at(Token::And)) {
            consume();
            Expr right = parse_not();
            left = make_bool(NodeKind::BoolAnd, left, right);
        }
        return left;
    }

    Expr parse_not() {
        if (at(Token::Not)) {
            consume();
            Expr operand = parse_cmp();
            auto n = std::make_shared<Node>();
            n->kind = NodeKind::Not;
            n->children = {operand};
            return n;
        }
        return parse_cmp();
    }

    Expr parse_cmp() {
        Expr left = parse_add();
        if (at(Token::Lt) || at(Token::LtEq) || at(Token::Gt) || at(Token::GtEq) ||
            at(Token::EqEq) || at(Token::NEq)) {
            std::string op = cur.text;
            consume();
            Expr right = parse_add();
            return make_cmpop(op, left, right);
        }
        return left;
    }

    Expr parse_add() {
        Expr left = parse_mul();
        while (at(Token::Plus) || at(Token::Minus)) {
            std::string op = cur.text;
            consume();
            Expr right = parse_mul();
            left = make_binop(op, left, right);
        }
        return left;
    }

    Expr parse_mul() {
        Expr left = parse_unary();
        while (at(Token::Star) || at(Token::Slash)) {
            std::string op = cur.text;
            consume();
            Expr right = parse_unary();
            left = make_binop(op, left, right);
        }
        return left;
    }

    Expr parse_unary() {
        if (at(Token::Minus)) {
            consume();
            Expr operand = parse_primary();
            auto n = std::make_shared<Node>();
            n->kind = NodeKind::Neg;
            n->children = {operand};
            return n;
        }
        return parse_primary();
    }

    Expr parse_primary() {
        if (at(Token::Num)) {
            auto n = std::make_shared<Node>();
            n->kind    = NodeKind::Lit;
            n->lit_val = cur.num;
            consume();
            return n;
        }

        if (at(Token::LParen)) {
            consume();
            Expr inner = parse_expr();
            if (!at(Token::RParen))
                throw std::invalid_argument("DSL: expected ')' in expression");
            consume();
            return inner;
        }

        if (at(Token::Id)) {
            std::string id = cur.text;
            consume();

            // pattern.X ?
            if (id == "pattern" && at(Token::Dot)) {
                consume();
                if (!at(Token::Id))
                    throw std::invalid_argument("DSL: expected pattern name after 'pattern.'");
                std::string pname = cur.text;
                consume();
                if (PATTERN_NAMES.find(pname) == PATTERN_NAMES.end())
                    throw std::invalid_argument(
                        "DSL: unknown pattern name '" + pname + "'");
                auto n = std::make_shared<Node>();
                n->kind = NodeKind::Pattern;
                n->name = pname;
                return n;
            }

            // Must be an allowed name
            if (allowed_names.find(id) == allowed_names.end()) {
                throw std::invalid_argument(
                    "DSL: name '" + id + "' is not declared");
            }
            auto n = std::make_shared<Node>();
            n->kind = NodeKind::Name;
            n->name = id;
            return n;
        }

        throw std::invalid_argument(
            "DSL: unexpected token '" + cur.text + "' in expression");
    }
};

/* =========================================================================
 * Public API
 * ========================================================================= */

Expr compile_expr(const std::string& src, const std::vector<std::string>& indicator_ids) {
    if (src.empty())
        throw std::invalid_argument("DSL: expression must be non-empty");
    Parser p(src, indicator_ids);
    Expr expr = p.parse_expr();
    if (p.cur.kind != Token::Eof)
        throw std::invalid_argument("DSL: unexpected token at end: '" + p.cur.text + "'");
    return expr;
}

/* =========================================================================
 * Evaluator
 * ========================================================================= */

static double eval_double(const Expr& node, const Env& env);

static double eval_double(const Expr& node, const Env& env) {
    switch (node->kind) {
        case NodeKind::Lit:
            return node->lit_val;

        case NodeKind::Name: {
            auto it = env.find(node->name);
            if (it == env.end()) return std::numeric_limits<double>::quiet_NaN();
            return it->second;
        }

        case NodeKind::Pattern: {
            // pattern values stored as 0.0/1.0 in env with prefix "pattern."
            auto it = env.find("pattern." + node->name);
            if (it != env.end()) return it->second;
            return 0.0;
        }

        case NodeKind::Neg:
            return -eval_double(node->children[0], env);

        case NodeKind::BinOp: {
            double l = eval_double(node->children[0], env);
            double r = eval_double(node->children[1], env);
            if (std::isnan(l) || std::isnan(r)) return std::numeric_limits<double>::quiet_NaN();
            if (node->name == "+") return l + r;
            if (node->name == "-") return l - r;
            if (node->name == "*") return l * r;
            if (node->name == "/") return (std::abs(r) > 1e-15) ? l / r : std::numeric_limits<double>::quiet_NaN();
            return std::numeric_limits<double>::quiet_NaN();
        }

        case NodeKind::CmpOp: {
            double l = eval_double(node->children[0], env);
            double r = eval_double(node->children[1], env);
            if (std::isnan(l) || std::isnan(r)) return 0.0;
            if (node->name == "<")  return l < r  ? 1.0 : 0.0;
            if (node->name == "<=") return l <= r ? 1.0 : 0.0;
            if (node->name == ">")  return l > r  ? 1.0 : 0.0;
            if (node->name == ">=") return l >= r ? 1.0 : 0.0;
            if (node->name == "==") return l == r ? 1.0 : 0.0;
            if (node->name == "!=") return l != r ? 1.0 : 0.0;
            return 0.0;
        }

        case NodeKind::BoolAnd: {
            double l = eval_double(node->children[0], env);
            if (std::isnan(l) || l == 0.0) return 0.0;
            double r = eval_double(node->children[1], env);
            return (std::isnan(r) || r == 0.0) ? 0.0 : 1.0;
        }

        case NodeKind::BoolOr: {
            double l = eval_double(node->children[0], env);
            if (!std::isnan(l) && l != 0.0) return 1.0;
            double r = eval_double(node->children[1], env);
            return (!std::isnan(r) && r != 0.0) ? 1.0 : 0.0;
        }

        case NodeKind::Not: {
            double v = eval_double(node->children[0], env);
            if (std::isnan(v)) return 1.0; // not NaN → treat as false → not false = true? No — treat as false.
            return v == 0.0 ? 1.0 : 0.0;
        }

        default:
            return std::numeric_limits<double>::quiet_NaN();
    }
}

bool eval_expr(const Expr& expr, const Env& env) {
    try {
        double v = eval_double(expr, env);
        if (std::isnan(v)) return false;
        return v != 0.0;
    } catch (...) {
        return false;
    }
}

} /* namespace algoforge::algo_gen::dsl */
